### PR TITLE
WebView support

### PR DESCRIPTION
**Pull request by galmacky**

### CODE REVIEW

**Comment by galmacky** [comment]
Design doc is here: https://docs.google.com/document/d/19cGz31lzCBdpbtSC92svXlhlhn68hrsVwSB7cfZt54o/edit

@Hexcles: this is my first CL for android_webview support.

File: tools/wpt/browser.py [hunk 1]
```diff
@@ -581,17 +581,16 @@ def version(self, binary=None, webdriver_binary=None):
         return m.group(1)
 
 
-class ChromeAndroid(Browser):
-    """Chrome-specific interface for Android.
+class ChromeAndroidBase(Browser):
```
Comment (Hexcles) [inline tools/wpt/browser.py]: I don't see the need for creating this base class. `AndroidWebview` can directly inherit `ChromeAndroid`; `AndroidWebview` can directly override any methods or properties that are different (including `product` and `requirements`).

This is fine, too, but we might want to mark this base class as abstract (e.g. `__metaclass__ = abc.ABCMeta`).

File: tools/wpt/browser.py [hunk 2]
```diff
@@ -629,6 +628,44 @@ def version(self, binary=None, webdriver_binary=None):
...
+        # For WebView, it is not trivial to change the WebView provider, so
+        # we will just grab whatever is available.
+        # https://chromium.googlesource.com/chromium/src/+/HEAD/android_webview/docs/channels.md
+        command = ['adb', 'shell', 'dumpsys', 'webviewupdate']
+        try:
+            output = call(*command)
+        except (subprocess.CalledProcessError, OSError):
+            self.logger.warning("Failed to call %s" % " ".join(command))
+            return None
+        m = re.search(r'^\s*Current WebView package \(name, version\): \((.*), ([0-9.]*)\)$',
```
Comment (Hexcles) [inline tools/wpt/browser.py]: Can we also use this for getting the version of WebView?

File: tools/wpt/browser.py [hunk 3]
```diff
@@ -643,10 +680,11 @@ def find_binary(self, venv_path=None, channel=None):
         raise NotImplementedError
 
     def find_webdriver(self, channel=None):
-        raise NotImplementedError
+        return find_executable("chromedriver")
```
Comment (Hexcles) [inline tools/wpt/browser.py]: `ChromeiOS` does not use `chromedriver` and is also in the prototype phase, so this (and `install_webdriver`) should `raise NotImplementedError`.

File: tools/wptrunner/wptrunner/browsers/android_webview.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
+import subprocess
+
+from .base import Browser, ExecutorBrowser, require_arg
+from .base import get_timeout_multiplier   # noqa: F401
+from ..webdriver_server import ChromeDriverServer
+from ..executors import executor_kwargs as base_executor_kwargs
+from ..executors.executorselenium import (SeleniumTestharnessExecutor,  # noqa: F401
```
Comment (Hexcles) [inline tools/wptrunner/wptrunner/browsers/android_webview.py]: `SeleniumTestharnessExecutor` is deprecated. I recently refactored `chrome_android.py` to use the new `WebDriverExecutor`. It's rather simple, just need to change the base classes: https://github.com/web-platform-tests/wpt/pull/18186/files#diff-ee29ae05c7ef60acfc8005fb7486d41d

File: tools/wpt/browser.py (hunk 1 above)
Comment (galmacky) [reply tools/wpt/browser.py]: It is just a matter of how we conceptualize this, but I'm worried that future developers of this file would just change ChromeAndroid without trying to check its impact on AndroidWebView class. Therefore, I prefer to keep the common logic into a base class. I'll make this abstract as you suggested.

File: tools/wpt/browser.py (hunk 2 above)
Comment (galmacky) [reply tools/wpt/browser.py]: Hmm... I don't think it's easy to retrofit it to the current logic of 1) getting the binary first and 2) get the version. And this separation allows us to provide a separate browser_binary, although I'm not sure if that's possible at all.

File: tools/wpt/browser.py (hunk 3 above)
Comment (galmacky) [reply tools/wpt/browser.py]: ChromeiOS is not going to inherit from ChromeAndroidBase, while AndroidWebView is going to and use this function.

Comment (Hexcles) [reply tools/wpt/browser.py]: No, it does not; it inherits from `Browser`, which isn't changed here, and we are not using or inheriting from `ChromeiOS` in `ChromeAndroidBase` or `AndroidWebView`, so I'm not sure why this needs to be changed.

File: tools/wptrunner/wptrunner/browsers/android_webview.py (hunk 4 above)
Comment (galmacky) [reply tools/wptrunner/wptrunner/browsers/android_webview.py]: Thanks for the pointer! Updated this file and now it's importing WebDriver*Executor and chrome_executor_kwargs.

**Comment by galmacky** [comment]
Updated the patch. @Hexcles, could you take another look? Thanks.

**Comment by galmacky** [comment]
@foolip Philip, could you take a look as well? Thanks.

**Comment by Hexcles** [comment]
Generally looks good except the remaining question regarding `ChromeiOS`

File: tools/wpt/browser.py (hunk 3 above)
Comment (galmacky) [reply tools/wpt/browser.py]: Oops, I didn't intend to change this part. ChromeiOS change has been removed now.

**Comment by galmacky** [comment]
Just a gentle ping.

**Review by Hexcles — APPROVED** [review]
LGTM

**Pull request merged**